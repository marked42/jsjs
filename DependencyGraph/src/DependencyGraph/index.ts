export * from './DependencyGraph'
export * from './DependencyGraphBuilder'
export * from './DependencyGraphNode'
