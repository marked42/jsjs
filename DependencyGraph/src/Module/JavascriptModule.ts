import { ModuleSpecifier } from './Module'
import { Module } from './Module'
import traverse from '@babel/traverse'
import { parse } from '@babel/parser'
import fs from 'fs'
import path from 'path'

export class JavascriptModule extends Module {
  constructor(
    private readonly _id: string,
    private readonly _filePath: string
  ) {
    super()
  }

  filePath(): string {
    return this._filePath
  }

  id(): string {
    return this._id
  }

  source(): string {
    return fs.readFileSync(this.filePath(), { encoding: 'utf-8' })
  }

  getSubModuleSpecifiers(): ModuleSpecifier[] {
    return getModuleSpecifiers(this.source())
  }
}

export function getModuleSpecifiers(source: string) {
  const ast = parse(source, {
    sourceType: 'unambiguous',
    plugins: ['typescript'],
  })

  const subModuleSpecifiers = [] as ModuleSpecifier[]
  if (!ast) {
    return subModuleSpecifiers
  }

  traverse(ast, {
    ImportDeclaration(path) {
      const moduleSpecifier = path.node.source.value

      subModuleSpecifiers.push(moduleSpecifier)
    },
  })

  return subModuleSpecifiers
}
