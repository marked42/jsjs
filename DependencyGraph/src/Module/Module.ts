export type ModuleSpecifier = string
export type ModuleId = string

export abstract class Module {
  abstract getSubModuleSpecifiers(): ModuleSpecifier[]

  abstract id(): ModuleId

  abstract source(): string

  abstract filePath(): string
}
