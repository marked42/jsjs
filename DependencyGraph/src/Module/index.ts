export * from './JavascriptModule'
export * from './Module'
