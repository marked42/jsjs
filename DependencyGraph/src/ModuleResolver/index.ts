export * from './ChainedModuleResolver'
export * from './JavascriptModuleResolver'
export * from './ModuleResolver'
