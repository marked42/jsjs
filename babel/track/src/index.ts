import { declare } from '@babel/helper-plugin-utils'
import * as t from '@babel/types'
import type { NodePath } from '@babel/core'
import template from '@babel/template'
// @ts-expect-error no typings
import { addDefault } from '@babel/helper-module-imports'

export const autoTrackPlugin = declare((api, options) => {
  api.assertVersion(7)

  return {
    name: 'track',

    visitor: {
      Program(path, state) {
        path.traverse({
          ImportDeclaration(decl) {
            if (decl.node.source.value === 'track') {
              const specifier = decl.get('specifiers.0')
              // if (specifier.isImportSpecifier()) {
              // }
            }
          },
        })
      },

      'ClassMethod|FunctionDeclaration|FunctionExpression|ArrowFunctionExpression'(
        path: NodePath<
          | t.ClassMethod
          | t.FunctionDeclaration
          | t.FunctionExpression
          | t.ArrowFunctionExpression
        >
      ) {
        const body = path.get('body')
        if (body.isBlockStatement()) {
          body.node.body.unshift(template.statement('pre();')())
        }
      },
    },
  }
})
