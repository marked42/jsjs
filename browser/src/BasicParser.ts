export class BasicParser {
  private index = 0
  constructor(private input: string) {}

  char() {
    return this.input[this.index]
  }

  next(len?: number) {
    len = len || 1
    this.index += len
  }

  eat(char?: string) {
    if (typeof char === 'string') {
      this.assert(char)
    }

    const text = char ? char : this.char()
    this.next(text ? text.length : 1)
    return text
  }

  assert(char: string) {
    if (!this.input.substring(this.index).startsWith(char)) {
      throw new Error(`unexpected char ${char}`)
    }
  }

  consumeWhitespace() {
    const whitespace = /[\t|\r|\n| ]/
    const startIndex = this.index
    while (whitespace.test(this.char())) {
      this.eat()
    }

    return this.input.substring(startIndex, this.index)
  }

  startsWith(input: string) {
    return this.input.substring(this.index).indexOf(input) === 0
  }

  end() {
    return this.index >= this.input.length
  }

  consumeCharsWhile(predicate: (char: string) => boolean) {
    let text = ''
    while (predicate(this.char()) && !this.end()) {
      text += this.eat()
    }

    return text
  }
}
