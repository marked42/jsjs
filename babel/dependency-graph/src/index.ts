import { ModuleSpecifier } from './Module'
import {
  ModuleResolver,
  ChainedModuleResolver,
  JavascriptModuleResolver,
} from './ModuleResolver'
import { DependencyGraphBuilder } from './DependencyGraph'

export function buildModuleDependencyGraph(entry: ModuleSpecifier) {
  const resolverChain: ModuleResolver[] = [new JavascriptModuleResolver()]
  const resolver = new ChainedModuleResolver(resolverChain)
  const builder = new DependencyGraphBuilder(resolver, entry)

  return builder.build()
}
