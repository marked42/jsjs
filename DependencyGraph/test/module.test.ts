import { getModuleSpecifiers } from '../src/Module'
import { buildModuleDependencyGraph } from '../src'
import path from 'path'

describe('getModuleSpecifiers', () => {
  it('should get module specifiers', () => {
    const source = `
	import 'mod1'
	import a from 'mod2'
	import { a1 as b1 } from 'mod3'
	import * as a2 from 'mod4'

	import a3, { a4 } from 'mod5'
	import a5, * as b3 from 'mod6'
	`
    const specifiers = getModuleSpecifiers(source)
    expect(specifiers).toMatchSnapshot()
  })

  it('should build module dependency graph by file path', () => {
    const graph = buildModuleDependencyGraph(
      path.join(__dirname, 'module-test', 'index.ts')
    )

    expect(graph).toMatchSnapshot()
  })

  it('should build module dependency graph by folder path', () => {
    const graph = buildModuleDependencyGraph(
      path.join(__dirname, 'module-test')
    )

    expect(graph).toMatchSnapshot()
  })
})
