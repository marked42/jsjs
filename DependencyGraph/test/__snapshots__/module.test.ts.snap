// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getModuleSpecifiers should build module dependency graph by file path 1`] = `
DependencyGraph {
  "allModules": Map {
    "test/module-test/index.ts" => JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/index.ts",
      "_id": "test/module-test/index.ts",
    },
    "math" => JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/math.ts",
      "_id": "math",
    },
    "data" => JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/data.ts",
      "_id": "data",
    },
  },
  "entry": DependencyGraphNode {
    "dependencies": Array [
      DependencyGraphNode {
        "dependencies": Array [],
        "value": JavascriptModule {
          "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/math.ts",
          "_id": "math",
        },
      },
      DependencyGraphNode {
        "dependencies": Array [],
        "value": JavascriptModule {
          "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/data.ts",
          "_id": "data",
        },
      },
    ],
    "value": JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/index.ts",
      "_id": "test/module-test/index.ts",
    },
  },
}
`;

exports[`getModuleSpecifiers should build module dependency graph by folder path 1`] = `
DependencyGraph {
  "allModules": Map {
    "test/module-test" => JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/index.ts",
      "_id": "test/module-test",
    },
    "math" => JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/math.ts",
      "_id": "math",
    },
    "data" => JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/data.ts",
      "_id": "data",
    },
  },
  "entry": DependencyGraphNode {
    "dependencies": Array [
      DependencyGraphNode {
        "dependencies": Array [],
        "value": JavascriptModule {
          "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/math.ts",
          "_id": "math",
        },
      },
      DependencyGraphNode {
        "dependencies": Array [],
        "value": JavascriptModule {
          "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/data.ts",
          "_id": "data",
        },
      },
    ],
    "value": JavascriptModule {
      "_filePath": "/Users/penghui/coding/jsjs/DependencyGraph/test/module-test/index.ts",
      "_id": "test/module-test",
    },
  },
}
`;

exports[`getModuleSpecifiers should get module specifiers 1`] = `
Array [
  "mod1",
  "mod2",
  "mod3",
  "mod4",
  "mod5",
  "mod6",
]
`;
